File: edu/yu/compilers/intermediate/symtable/SymTableStack.java
/**
 * <h1>SymTableStack</h1>
 * <p>The symbol table stack.</p>
 * <p>Adapted from:</p>
 * <p>Copyright (c) 2020 by Ronald Mak</p>
 * <p>For instructional purposes only.  No warranties.</p>
 */

package edu.yu.compilers.intermediate.symtable;

import edu.yu.compilers.intermediate.symtable.SymTableEntry.Kind;

import java.util.ArrayList;

public class SymTableStack extends ArrayList<SymTable> {
    private static final long serialVersionUID = 0L;

    private int currentNestingLevel;  // current scope nesting level
    private SymTableEntry programId;    // entry for the main program id

    /**
     * Constructor.
     */
    public SymTableStack() {
        this.currentNestingLevel = 0;
        add(new SymTable(currentNestingLevel));
    }

    /**
     * Getter.
     *
     * @return the current nesting level.
     */
    public int getCurrentNestingLevel() {
        return currentNestingLevel;
    }

    /**
     * Getter.
     *
     * @return the symbol table entry for the main program identifier.
     */
    public SymTableEntry getProgramId() {
        return programId;
    }

    /**
     * Setter.
     *
     * @param id the symbol table entry for the main program identifier.
     */
    public void setProgramId(SymTableEntry id) {
        this.programId = id;
    }

    /**
     * Return the local symbol table which is at the top of the stack.
     *
     * @return the local symbol table.
     */
    public SymTable getLocalSymTable() {
        return get(currentNestingLevel);
    }

    /**
     * Push a new symbol table onto the symbol table stack.
     *
     * @return the pushed symbol table.
     */
    public SymTable push() {
        SymTable symTable = new SymTable(++currentNestingLevel);
        add(symTable);

        return symTable;
    }

    /**
     * Push a symbol table onto the symbol table stack.
     *
     * @return the pushed symbol table.
     */
    public SymTable push(SymTable symTable) {
        ++currentNestingLevel;
        add(symTable);

        return symTable;
    }

    /**
     * Pop a symbol table off the symbol table stack.
     *
     * @return the popped symbol table.
     */
    public SymTable pop() {
        SymTable symTable = get(currentNestingLevel);
        remove(currentNestingLevel--);

        return symTable;
    }

    /**
     * Create and enter a new entry into the local symbol table.
     *
     * @param name the name of the entry.
     * @param kind what kind of entry.
     * @return the new entry.
     */
    public SymTableEntry enterLocal(String name, Kind kind) {
        return get(currentNestingLevel).enter(name, kind);
    }

    /**
     * Look up an existing symbol table entry in the local symbol table.
     *
     * @param name the name of the entry.
     * @return the entry, or null if it does not exist.
     */
    public SymTableEntry lookupLocal(String name) {
        return get(currentNestingLevel).lookup(name);
    }

    /**
     * Look up an existing symbol table entry throughout the stack.
     *
     * @param name the name of the entry.
     * @return the entry, or null if it does not exist.
     */
    public SymTableEntry lookup(String name) {
        SymTableEntry foundEntry = null;

        // Search the current and enclosing scopes.
        for (int i = currentNestingLevel; (i >= 0) && (foundEntry == null); --i) {
            foundEntry = get(i).lookup(name);
        }

        return foundEntry;
    }
}

File: edu/yu/compilers/intermediate/symtable/SymTableStack.py
from multipledispatch import dispatch
from edu.yu.compilers.intermediate.symtable.SymTable import SymTable
from edu.yu.compilers.intermediate.symtable.SymTableEntry import SymTableEntry


class SymTableStack(list):
    serialVersionUID = 0

    def __init__(self):
        super().__init__()
        self.currentNestingLevel = 0
        # change all below super() to self if doesnt work
        self.append(SymTable(self.currentNestingLevel))  # ChatGPT used self instead of super()
        self.programId = None

    def getCurrentNestingLevel(self) -> int:  # chatGPT used - > <ret type> :
        return self.currentNestingLevel

    def getProgramId(self):
        return self.programId

    def setProgramId(self, _id: SymTableEntry):
        self.programId = _id

    def getLocalSymTable(self):
        return self[self.currentNestingLevel]

    # ChatGPT Did not seem to properly override the methods in Python

    @dispatch()  # Not sure if this will work
    def push(self):
        self.currentNestingLevel += 1
        symTable = SymTable(self.currentNestingLevel)
        self.append(symTable)
        return symTable

    @dispatch(SymTable)  # FIXME
    def push(self, symTable):
        self.currentNestingLevel += 1
        self.append(symTable)

        return symTable

    def _pop(self):  # TODO no issues?
        symTable = self[self.currentNestingLevel]
        self.pop(self.currentNestingLevel)
        self.currentNestingLevel -= 1
        # We differ slightly but may be functionally identical - I remove by index
        # ChatGPT simply pops - but I think same effect is achieved.

        return symTable

    def enterLocal(self, name, kind):
        return self[self.currentNestingLevel].enter(name, kind)  # This is correct

    def lookupLocal(self, name):
        return self[self.currentNestingLevel].lookup(name)  # This is correct

    def lookup(self, name):
        foundEntry = None

        # Search the current and enclosing scopes.
        # i = self.currentNestingLevel
        # while (i >= 0) and (foundEntry is None):
        #     foundEntry = self[i].lookup(name)
        #     i -= 1 # i made decrememnt mistake i think
        #     # it also added a break statement
        # NO WAAY CRAZY
        # ChatGPT's loop - very nice
        for i in range(self.currentNestingLevel, -1, -1):
            foundEntry = self[i].lookup(name)
            if foundEntry:
                break

        return foundEntry
